Realm/Tests/KVOTests.mm

```diff
@@ -0,0 +1,1278 @@
+    XCTAssertNoThrow([obj removeObserver:self forKeyPath:@"int32Col"]);
+}
+
+- (void)testSimple {
+    KVOObject *obj = [self createObject];
+    {
+        KVORecorder r(self, obj, @"int32Col");
+        obj.int32Col = 10;
+        AssertChanged(r, @2, @10);
+    }
+    {
+        KVORecorder r(self, obj, @"int32Col");
+        obj.int32Col = 1;
+        AssertChanged(r, @10, @1);
+    }
+}
+
+- (void)testSelfAssignmentNotifies {
+    KVOObject *obj = [self createObject];
+    {
+        KVORecorder r(self, obj, @"int32Col");
+        obj.int32Col = obj.int32Col;
+        AssertChanged(r, @2, @2);
+    }
+}
+
+- (void)testMultipleObserversAreNotified {
+    KVOObject *obj = [self createObject];
+    {
+        KVORecorder r1(self, obj, @"int32Col");
+        KVORecorder r2(self, obj, @"int32Col");
+        KVORecorder r3(self, obj, @"int32Col");
+        obj.int32Col = 10;
+        AssertChanged(r1, @2, @10);
+        AssertChanged(r2, @2, @10);
+        AssertChanged(r3, @2, @10);
+    }
+}
+
+- (void)testOnlyObserversForTheCorrectPropertyAreNotified {
+    KVOObject *obj = [self createObject];
+    {
+        KVORecorder r16(self, obj, @"int16Col");
+        KVORecorder r32(self, obj, @"int32Col");
+        KVORecorder r64(self, obj, @"int64Col");
+
+        obj.int16Col = 2;
+        AssertChanged(r16, @1, @2);
+        XCTAssertTrue(r16.empty());
+        XCTAssertTrue(r32.empty());
+        XCTAssertTrue(r64.empty());
+
+        obj.int32Col = 2;
+        AssertChanged(r32, @2, @2);
+        XCTAssertTrue(r16.empty());
+        XCTAssertTrue(r32.empty());
+        XCTAssertTrue(r64.empty());
+
+        obj.int64Col = 2;
+        AssertChanged(r64, @3, @2);
+        XCTAssertTrue(r16.empty());
+        XCTAssertTrue(r32.empty());
+        XCTAssertTrue(r64.empty());
+    }
+}
+
+- (void)testMultipleChangesWithSingleObserver {
+    KVOObject *obj = [self createObject];
+    KVORecorder r(self, obj, @"int32Col");
+
+    obj.int32Col = 1;
+    obj.int32Col = 2;
+    obj.int32Col = 3;
+    obj.int32Col = 3;
+
+    if (self.collapsesNotifications) {
+        AssertChanged(r, @2, @3);
+    }
+    else {
+        AssertChanged(r, @2, @1);
+        AssertChanged(r, @1, @2);
+        AssertChanged(r, @2, @3);
+        AssertChanged(r, @3, @3);
+    }
+}
+
+- (void)testOnlyObserversForTheCorrectObjectAreNotified {
+    KVOObject *obj1 = [self createObject];
+    KVOObject *obj2 = [self createObject];
+
+    KVORecorder r1(self, obj1, @"int32Col");
+    KVORecorder r2(self, obj2, @"int32Col");
+
+    obj1.int32Col = 10;
+    AssertChanged(r1, @2, @10);
+    XCTAssertEqual(0U, r2.size());
+
+    obj2.int32Col = 5;
+    AssertChanged(r2, @2, @5);
+}
+
+- (void)testOptionsInitial {
+    KVOObject *obj = [self createObject];
+
+    {
+        KVORecorder r(self, obj, @"int32Col", 0);
+        XCTAssertEqual(0U, r.size());
+    }
+    {
+        KVORecorder r(self, obj, @"int32Col", NSKeyValueObservingOptionInitial);
+        r.pop_front();
```

Comment:
that there is exactly one notification (pop_front asserts that there's at least one and KVORecorder's destructor asserts that all notifications were consumed)